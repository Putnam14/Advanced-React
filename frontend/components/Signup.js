import React, { Component } from 'react';
import { Mutation, ApolloConsumer } from 'react-apollo';
import { SIGNUP_MUTATION, CURRENT_USER_QUERY } from '../queries/queries';
import Form from './styles/Form';
import Error from './ErrorMessage';

class Signup extends Component {
  state = {
    email: `wesbos@gmail.com`,
    name: 'Wes Bos',
    password: 'wes',
  };

  saveToState = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  render() {
    return (
      <ApolloConsumer>
        {client => (
          <Mutation mutation={SIGNUP_MUTATION} variables={this.state}>
            {(signup, { loading, error }) => (
              <Form
                onSubmit={async e => {
                  e.preventDefault();
                  const res = await signup();
                  localStorage.setItem('token', res.data.signup.token);
                  client.query({ query: CURRENT_USER_QUERY, fetchPolicy: 'network-only' });
                }}
              >
                <fieldset disabled={loading} aria-busy={loading}>
                  <Error error={error} />
                  <h2>Sign Up for an Account</h2>
                  <label htmlFor="email">
                    Email
                    <input
                      value={this.state.email}
                      onChange={this.saveToState}
                      name="email"
                      type="text"
                      placeholder="email"
                    />
                  </label>

                  <label htmlFor="name">
                    Name
                    <input
                      type="text"
                      name="name"
                      placeholder="name"
                      value={this.state.name}
                      onChange={this.saveToState}
                    />
                  </label>

                  <label htmlFor="signupPassword">
                    Password
                    <input
                      type="password"
                      name="password"
                      id="signupPassword"
                      className="password"
                      placeholder="password"
                      value={this.state.password}
                      onChange={this.saveToState}
                    />
                  </label>

                  <button type="submit">Submit</button>
                </fieldset>
              </Form>
            )}
          </Mutation>
        )}
      </ApolloConsumer>
    );
  }
}

export default Signup;
