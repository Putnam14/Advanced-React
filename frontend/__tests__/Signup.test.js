import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Signup from '../components/Signup';
import mountOptions from '../lib/testUtils';

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: {
      name,
      value,
    },
  });
}

describe('<Signup/>', () => {
  beforeAll(() => {
    // mock localStorage
    global.localStorage = {
      setItem: jest.fn(),
    };
  });

  it('renders and matches snapshot', () => {
    const wrapper = mount(<Signup />, mountOptions);
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('calls mutation with correct data', async () => {
    const wrapper = mount(<Signup />, mountOptions);
    const mutation = wrapper.find('Mutation').instance();
    mutation.client.mutate = jest.fn().mockResolvedValue({
      data: { signup: { token: 'omg123' } },
    });

    const expected = {
      name: 'Randy',
      email: 'Randy@randymail.net',
      password: 'pupsrock',
    };

    type(wrapper, 'name', expected.name);
    type(wrapper, 'email', expected.email);
    type(wrapper, 'password', expected.password);

    wrapper.simulate('submit');

    expect(mutation.client.mutate).toHaveBeenCalled();
    expect(mutation.client.mutate).toHaveBeenCalledWithVariables(expected);
    await wait();
    expect(localStorage.setItem).toHaveBeenCalledWith('token', 'omg123');
  });
});
