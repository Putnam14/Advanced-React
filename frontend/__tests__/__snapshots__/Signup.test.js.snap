// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="sc-bdVaJa LHejR"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <DisplayError />
    <h2>
      Sign Up for an Account
    </h2>
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="text"
        value="wesbos@gmail.com"
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="name"
        type="text"
        value="Wes Bos"
      />
    </label>
    <label
      htmlFor="signupPassword"
    >
      Password
      <input
        className="password"
        id="signupPassword"
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value="wes"
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
